clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/consumer/ConsumerBounceTest.java

```diff
@@ -0,0 +1,807 @@
+            try {
+                response = IntegrationTestUtils.connectAndReceive(request, clusterInstance.anyBrokerSocketServer().boundPort(new ListenerName("EXTERNAL")));
+            } catch (IOException e) {
+                return false;
+            }
+
+            if (response.hasError())
+                return false;
+            for (String group : groups)
+                if (response.coordinatorByKey(group).isEmpty())
+                    return false;
+                else
+                    nodes.add(response.coordinatorByKey(group).get().nodeId());
+            return true;
+        }, "Failed to find coordinator for group " + groups);
+        return nodes;
+    }
+
+    @ClusterTest
+    public void testClassicConsumerReceivesFatalExceptionWhenGroupPassesMaxSize() throws Exception {
+        testConsumerReceivesFatalExceptionWhenGroupPassesMaxSize(GroupProtocol.CLASSIC);
+    }
+
+    @ClusterTest
+    public void testAsyncConsumerReceivesFatalExceptionWhenGroupPassesMaxSize() throws Exception {
+        testConsumerReceivesFatalExceptionWhenGroupPassesMaxSize(GroupProtocol.CONSUMER);
+    }
+
+    private void testConsumerReceivesFatalExceptionWhenGroupPassesMaxSize(GroupProtocol groupProtocol) throws Exception {
+        String group = "fatal-exception-test";
+        String topic = "fatal-exception-test";
+
+        Map<String, String> consumerConfig = new HashMap<>();
+        int numPartition = Integer.parseInt(MAX_GROUP_SIZE);
+
+        consumerConfig.put(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG, "60000");
+        if (groupProtocol.equals(GroupProtocol.CLASSIC)) {
+            consumerConfig.put(ConsumerConfig.HEARTBEAT_INTERVAL_MS_CONFIG, "1000");
+        }
+        consumerConfig.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false");
+
+        clusterInstance.createTopic(topic, Integer.parseInt(MAX_GROUP_SIZE), (short) BROKER_COUNT);
+        Set<TopicPartition> partitions = new HashSet<>();
+        for (int i = 0; i < Integer.parseInt(MAX_GROUP_SIZE); ++i)
+            partitions.add(new TopicPartition(topic, i));
+
+        addConsumersToGroupAndWaitForGroupAssignment(
+                Integer.parseInt(MAX_GROUP_SIZE),
+                List.of(topic),
+                partitions,
+                group,
+                consumerConfig
+        );
+
+        addConsumersToGroup(
+                1,
+                List.of(topic),
+                group,
+                consumerConfig
+        );
+
+        ConsumerAssignmentPoller rejectedConsumer = consumerPollers.get(consumerPollers.size() - 1);
+        consumerPollers.remove(consumerPollers.size() - 1);
+
+        TestUtils.waitForCondition(
+                () -> rejectedConsumer.getThrownException().isPresent(),
+                "Extra consumer did not throw an exception"
+        );
+
+        assertInstanceOf(GroupMaxSizeReachedException.class, rejectedConsumer.getThrownException().get());
+
+        // assert group continues to live
```

Comment:
Could you please enrich the comment? We expect the records to be distributed across all partitions.. For another, perhaps we can simplify the code.
```java
        var data = "data".getBytes(StandardCharsets.UTF_8);
        try (Producer<byte[], byte[]> producer = clusterInstance.producer()) {
            IntStream.range(0, numPartition * 100).forEach(index ->
                producer.send(new ProducerRecord<>(topic, index % numPartition, data, data)));
        }
```